network/default.go

```diff
@@ -647,41 +539,23 @@ func (n *network) processCtrlChan(client transport.Client, listener tunnel.Liste
 				if pbRtrAdvert.Id == n.options.Id {
 					continue
 				}
-				// loookup advertising node in our neighbourhood
-				n.RLock()
-				log.Debugf("Network received advert message from: %s", pbRtrAdvert.Id)
-				advertNode, ok := n.neighbours[pbRtrAdvert.Id]
-				if !ok {
-					// advertising node has not been registered as our neighbour, yet
-					// let's add it to the map of our neighbours
-					advertNode = &node{
-						id:         pbRtrAdvert.Id,
-						neighbours: make(map[string]*node),
-						lastSeen:   now,
-					}
-					n.neighbours[pbRtrAdvert.Id] = advertNode
-					// send a solicit message when discovering a new node
-					if err := n.sendMsg("solicit", NetworkChannel); err != nil {
-						log.Debugf("Network failed to send solicit message: %s", err)
-					}
+				log.Debugf("Network received advert message with %d events from: %s", len(pbRtrAdvert.Events), pbRtrAdvert.Id)
+				// loookup advertising node in our peer topology
+				advertNode := n.node.GetPeerNode(pbRtrAdvert.Id)
+				if advertNode == nil {
+					// if we can't find the node in our topology (MaxDepth) we skipp prcessing adverts
+					log.Debugf("Network skipping advert message from unknown peer: %s", pbRtrAdvert.Id)
+					continue
 				}
-				n.RUnlock()
 
 				var events []*router.Event
 				for _, event := range pbRtrAdvert.Events {
-					// set the address of the advertising node
-					// we know Route.Gateway is the address of advertNode
-					// NOTE: this is true only when advertNode had not been registered
-					// as our neighbour when we received the advert from it
-					if advertNode.address == "" {
-						advertNode.address = event.Route.Gateway
-					}
-					// if advertising node id is not the same as Route.Router
 					// we know the advertising node is not the origin of the route
-					if advertNode.id != event.Route.Router {
-						// if the origin router is not in the advertising node neighbourhood
+					if pbRtrAdvert.Id != event.Route.Router {
+						// if the origin router is not the advertising node peer
 						// we can't rule out potential routing loops so we bail here
-						if _, ok := advertNode.neighbours[event.Route.Router]; !ok {
+						if !advertNode.HasPeer(event.Route.Router) {
```

Comment:
This will throw away routes that are more than one hop away.

If the chain is (A - B - C - D) then A will throw away routes from B that were originated at D. It will only accept routes from B-C and nothing else.

Reply:
```diff --git a/network/default.go b/network/default.go
index c981f1d..605fa01 100644
--- a/network/default.go
+++ b/network/default.go
@@ -554,7 +554,7 @@ func (n *network) processCtrlChan(client transport.Client, listener tunnel.Liste
                                        if pbRtrAdvert.Id != event.Route.Router {
                                                // if the origin router is not the advertising node peer
                                                // we can't rule out potential routing loops so we bail here
-                                               if !advertNode.HasPeer(event.Route.Router) {
+                                               if peer := advertNode.GetPeerNode(event.Route.Router); peer == nil {
                                                        log.Debugf("Network skipping advert message from peer: %s", pbRtrAdvert.Id)
                                                        continue
                                                }
```